File: crypto_screener/coinglass.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

import httpx

from .providers import ProviderError


@dataclass(frozen=True)
class CoinGlassClient:
    api_key: str
    base_url: str = "https://open-api-v4.coinglass.com"
    timeout_seconds: float = 12
    user_agent: str = "codex-crypto-screener/0.2"

    def get_json(self, path: str, params: dict[str, Any] | None = None) -> Any:
        if not self.api_key:
            raise ProviderError("CoinGlass API key is not set")

        url = self.base_url.rstrip("/") + "/" + path.lstrip("/")
        try:
            response = httpx.get(
                url,
                params=params,
                headers={
                    "Accept": "application/json",
                    "Content-Type": "application/json",
                    "CG-API-KEY": self.api_key,
                    "User-Agent": self.user_agent,
                },
                timeout=self.timeout_seconds,
            )
            response.raise_for_status()
            payload = response.json()
        except httpx.HTTPStatusError as exc:
            body = exc.response.text[:500]
            raise ProviderError(f"{path} returned HTTP {exc.response.status_code}: {body}") from exc
        except httpx.TimeoutException as exc:
            raise ProviderError(f"{path} timed out after {self.timeout_seconds}s") from exc
        except httpx.HTTPError as exc:
            raise ProviderError(f"{path} failed: {exc}") from exc
        except json.JSONDecodeError as exc:
            raise ProviderError(f"{path} returned invalid JSON") from exc

        if not isinstance(payload, dict):
            raise ProviderError(f"{path} returned non-object JSON payload")

        code = str(payload.get("code", "0"))
        if code not in {"0", "200"}:
            raise ProviderError(f"{path} returned code {code}: {payload.get('msg')}")
        return payload.get("data")

    def futures_pairs_markets(self, symbol: str) -> list[dict[str, Any]]:
        data = self.get_json("/api/futures/pairs-markets", {"symbol": symbol})
        return data if isinstance(data, list) else []

    def price_history(
        self,
        exchange: str,
        symbol: str,
        interval: str,
        limit: int,
        start_time: int | None = None,
        end_time: int | None = None,
    ) -> list[dict[str, Any]]:
        params = {
            "exchange": exchange,
            "symbol": symbol,
            "interval": interval,
            "limit": limit,
        }
        if start_time is not None:
            params["start_time"] = start_time
        if end_time is not None:
            params["end_time"] = end_time
        data = self.get_json("/api/futures/price/history", params)
        return data if isinstance(data, list) else []

    def supported_coins(self) -> list[str]:
        data = self.get_json("/api/futures/supported-coins")
        return [str(item) for item in data] if isinstance(data, list) else []

    def supported_exchange_pairs(self, exchange: str | None = None) -> dict[str, list[dict[str, Any]]]:
        params = {"exchange": exchange} if exchange else None
        data = self.get_json("/api/futures/supported-exchange-pairs", params)
        if not isinstance(data, dict):
            return {}
        return {str(exchange_name): pairs for exchange_name, pairs in data.items() if isinstance(pairs, list)}

    def open_interest_aggregated_history(
        self,
        symbol: str,
        interval: str,
        limit: int,
        unit: str = "usd",
        start_time: int | None = None,
        end_time: int | None = None,
    ) -> list[dict[str, Any]]:
        params = {"symbol": symbol, "interval": interval, "limit": limit, "unit": unit}
        if start_time is not None:
            params["start_time"] = start_time
        if end_time is not None:
            params["end_time"] = end_time
        data = self.get_json("/api/futures/open-interest/aggregated-history", params)
        return data if isinstance(data, list) else []

    def funding_oi_weight_history(
        self,
        symbol: str,
        interval: str,
        limit: int,
        start_time: int | None = None,
        end_time: int | None = None,
    ) -> list[dict[str, Any]]:
        params = {"symbol": symbol, "interval": interval, "limit": limit}
        if start_time is not None:
            params["start_time"] = start_time
        if end_time is not None:
            params["end_time"] = end_time
        data = self.get_json("/api/futures/funding-rate/oi-weight-history", params)
        return data if isinstance(data, list) else []

    def liquidation_aggregated_history(
        self,
        exchange_list: list[str],
        symbol: str,
        interval: str,
        limit: int,
        start_time: int | None = None,
        end_time: int | None = None,
    ) -> list[dict[str, Any]]:
        params = {
            "exchange_list": ",".join(exchange_list),
            "symbol": symbol,
            "interval": interval,
            "limit": limit,
        }
        if start_time is not None:
            params["start_time"] = start_time
        if end_time is not None:
            params["end_time"] = end_time
        data = self.get_json("/api/futures/liquidation/aggregated-history", params)
        return data if isinstance(data, list) else []

    def aggregated_taker_buy_sell_history(
        self,
        exchange_list: list[str],
        symbol: str,
        interval: str,
        limit: int,
        unit: str = "usd",
        start_time: int | None = None,
        end_time: int | None = None,
    ) -> list[dict[str, Any]]:
        params = {
            "exchange_list": ",".join(exchange_list),
            "symbol": symbol,
            "interval": interval,
            "limit": limit,
            "unit": unit,
        }
        if start_time is not None:
            params["start_time"] = start_time
        if end_time is not None:
            params["end_time"] = end_time
        data = self.get_json("/api/futures/aggregated-taker-buy-sell-volume/history", params)
        return data if isinstance(data, list) else []

File: apps/api/src/providers/coinglass.ts
import { ProviderError } from './errors.js';
import { buildUrl, fetchWithRetry429, parseJsonResponse, RequestPacer } from './http.js';

// Loosely typed on purpose: fields are read defensively via `toFloat`, not exhaustively modeled.
export type CoinGlassPair = Record<string, unknown>;
export type CoinGlassHistoryRow = Record<string, unknown>;

export interface CoinGlassClient {
  supportedExchangePairs(exchange?: string): Promise<Record<string, CoinGlassPair[]>>;
  futuresPairsMarkets(symbol: string): Promise<CoinGlassPair[]>;
  priceHistory(
    exchange: string,
    symbol: string,
    interval: string,
    limit: number,
    startTime?: number,
    endTime?: number,
  ): Promise<CoinGlassHistoryRow[]>;
  openInterestAggregatedHistory(
    symbol: string,
    interval: string,
    limit: number,
    unit?: string,
    startTime?: number,
    endTime?: number,
  ): Promise<CoinGlassHistoryRow[]>;
  fundingOiWeightHistory(
    symbol: string,
    interval: string,
    limit: number,
    startTime?: number,
    endTime?: number,
  ): Promise<CoinGlassHistoryRow[]>;
  liquidationAggregatedHistory(
    exchangeList: string[],
    symbol: string,
    interval: string,
    limit: number,
    startTime?: number,
    endTime?: number,
  ): Promise<CoinGlassHistoryRow[]>;
  aggregatedTakerBuySellHistory(
    exchangeList: string[],
    symbol: string,
    interval: string,
    limit: number,
    unit?: string,
    startTime?: number,
    endTime?: number,
  ): Promise<CoinGlassHistoryRow[]>;
  globalLongShortAccountRatioHistory(
    exchange: string,
    symbol: string,
    interval: string,
    limit: number,
    startTime?: number,
    endTime?: number,
  ): Promise<CoinGlassHistoryRow[]>;
  topLongShortAccountRatioHistory(
    exchange: string,
    symbol: string,
    interval: string,
    limit: number,
    startTime?: number,
    endTime?: number,
  ): Promise<CoinGlassHistoryRow[]>;
  topLongShortPositionRatioHistory(
    exchange: string,
    symbol: string,
    interval: string,
    limit: number,
    startTime?: number,
    endTime?: number,
  ): Promise<CoinGlassHistoryRow[]>;
}

export interface CoinGlassClientOptions {
  apiKey: string;
  baseUrl?: string;
  timeoutSeconds?: number;
  userAgent?: string;
  retry429?: boolean;
  retry429InitialDelaySeconds?: number;
  retry429MaxDelaySeconds?: number;
  retry429JitterSeconds?: number;
  retry429MaxAttempts?: number;
  // Spacing between outgoing requests (RequestPacer, http.ts), applied at a single choke point in
  // getJson() below -- not per-loop sleeps scattered through collector.ts/enrichment.ts anymore.
  // Defaults to 0 (no pacing) so a client built without an explicit value -- e.g. most unit tests
  // that construct this class directly -- never waits; production always passes
  // providers.coinglass.request_delay_seconds explicitly (collector.ts).
  requestDelaySeconds?: number;
}

type QueryParams = Record<string, string | number | boolean | undefined>;

export class CoinGlassHttpClient implements CoinGlassClient {
  private readonly apiKey: string;
  private readonly baseUrl: string;
  private readonly timeoutSeconds: number;
  private readonly userAgent: string;
  private readonly retry429: boolean;
  private readonly retry429InitialDelaySeconds: number;
  private readonly retry429MaxDelaySeconds: number;
  private readonly retry429JitterSeconds: number;
  private readonly retry429MaxAttempts: number;
  private readonly pacer: RequestPacer;

  constructor(options: CoinGlassClientOptions) {
    this.apiKey = options.apiKey;
    this.baseUrl = options.baseUrl ?? 'https://open-api-v4.coinglass.com';
    this.timeoutSeconds = options.timeoutSeconds ?? 12;
    this.userAgent = options.userAgent ?? 'codex-crypto-screener/0.2';
    this.retry429 = options.retry429 ?? true;
    this.retry429InitialDelaySeconds = options.retry429InitialDelaySeconds ?? 10;
    this.retry429MaxDelaySeconds = options.retry429MaxDelaySeconds ?? 120;
    this.retry429JitterSeconds = options.retry429JitterSeconds ?? 5;
    this.retry429MaxAttempts = options.retry429MaxAttempts ?? 3;
    this.pacer = new RequestPacer(options.requestDelaySeconds ?? 0);
  }

  private async getJson(path: string, params?: QueryParams): Promise<unknown> {
    if (!this.apiKey) {
      throw new ProviderError('CoinGlass API key is not set');
    }

    // Single choke point: every outgoing request -- supportedExchangePairs, futuresPairsMarkets,
    // and every history method -- funnels through getJson, so pacing here covers all of them.
    // Applies once per logical request, not per 429-retry attempt: fetchWithRetry429 below already
    // spaces its own retries via backoff, and pacing each attempt too would double-wait.
    await this.pacer.pace();

    const url = buildUrl(this.baseUrl, path, params);
    const headers: Record<string, string> = {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      'CG-API-KEY': this.apiKey,
      'User-Agent': this.userAgent,
    };

    const response = await fetchWithRetry429(
      url,
      { timeoutSeconds: this.timeoutSeconds, headers },
      {
        enabled: this.retry429,
        initialDelaySeconds: this.retry429InitialDelaySeconds,
        maxDelaySeconds: this.retry429MaxDelaySeconds,
        jitterSeconds: this.retry429JitterSeconds,
        maxAttempts: this.retry429MaxAttempts,
      },
    );
    const payload = parseJsonResponse(path, response);

    if (typeof payload !== 'object' || payload === null || Array.isArray(payload)) {
      throw new ProviderError(`${path} returned non-object JSON payload`);
    }

    const record = payload as Record<string, unknown>;
    const code = String(record.code ?? '0');
    if (code !== '0' && code !== '200') {
      throw new ProviderError(`${path} returned code ${code}: ${String(record.msg)}`);
    }
    return record.data;
  }

  // Coerces a non-array payload to `[]` rather than throwing.
  private async getJsonArray(
    path: string,
    params?: QueryParams,
  ): Promise<Record<string, unknown>[]> {
    const data = await this.getJson(path, params);
    return Array.isArray(data) ? (data as Record<string, unknown>[]) : [];
  }

  async supportedExchangePairs(exchange?: string): Promise<Record<string, CoinGlassPair[]>> {
    const data = await this.getJson(
      '/api/futures/supported-exchange-pairs',
      exchange ? { exchange } : undefined,
    );
    if (typeof data !== 'object' || data === null || Array.isArray(data)) {
      return {};
    }
    const result: Record<string, CoinGlassPair[]> = {};
    for (const [exchangeName, pairs] of Object.entries(data as Record<string, unknown>)) {
      if (Array.isArray(pairs)) {
        result[exchangeName] = pairs as CoinGlassPair[];
      }
    }
    return result;
  }

  async futuresPairsMarkets(symbol: string): Promise<CoinGlassPair[]> {
    return this.getJsonArray('/api/futures/pairs-markets', { symbol });
  }

  async priceHistory(
    exchange: string,
    symbol: string,
    interval: string,
    limit: number,
    startTime?: number,
    endTime?: number,
  ): Promise<CoinGlassHistoryRow[]> {
    return this.getJsonArray('/api/futures/price/history', {
      exchange,
      symbol,
      interval,
      limit,
      start_time: startTime,
      end_time: endTime,
    });
  }

  async openInterestAggregatedHistory(
    symbol: string,
    interval: string,
    limit: number,
    unit = 'usd',
    startTime?: number,
    endTime?: number,
  ): Promise<CoinGlassHistoryRow[]> {
    return this.getJsonArray('/api/futures/open-interest/aggregated-history', {
      symbol,
      interval,
      limit,
      unit,
      start_time: startTime,
      end_time: endTime,
    });
  }

  async fundingOiWeightHistory(
    symbol: string,
    interval: string,
    limit: number,
    startTime?: number,
    endTime?: number,
  ): Promise<CoinGlassHistoryRow[]> {
    return this.getJsonArray('/api/futures/funding-rate/oi-weight-history', {
      symbol,
      interval,
      limit,
      start_time: startTime,
      end_time: endTime,
    });
  }

  async liquidationAggregatedHistory(
    exchangeList: string[],
    symbol: string,
    interval: string,
    limit: number,
    startTime?: number,
    endTime?: number,
  ): Promise<CoinGlassHistoryRow[]> {
    return this.getJsonArray('/api/futures/liquidation/aggregated-history', {
      exchange_list: exchangeList.join(','),
      symbol,
      interval,
      limit,
      start_time: startTime,
      end_time: endTime,
    });
  }

  async aggregatedTakerBuySellHistory(
    exchangeList: string[],
    symbol: string,
    interval: string,
    limit: number,
    unit = 'usd',
    startTime?: number,
    endTime?: number,
  ): Promise<CoinGlassHistoryRow[]> {
    return this.getJsonArray('/api/futures/aggregated-taker-buy-sell-volume/history', {
      exchange_list: exchangeList.join(','),
      symbol,
      interval,
      limit,
      unit,
      start_time: startTime,
      end_time: endTime,
    });
  }

  async globalLongShortAccountRatioHistory(
    exchange: string,
    symbol: string,
    interval: string,
    limit: number,
    startTime?: number,
    endTime?: number,
  ): Promise<CoinGlassHistoryRow[]> {
    return this.getJsonArray('/api/futures/global-long-short-account-ratio/history', {
      exchange,
      symbol,
      interval,
      limit,
      start_time: startTime,
      end_time: endTime,
    });
  }

  async topLongShortAccountRatioHistory(
    exchange: string,
    symbol: string,
    interval: string,
    limit: number,
    startTime?: number,
    endTime?: number,
  ): Promise<CoinGlassHistoryRow[]> {
    return this.getJsonArray('/api/futures/top-long-short-account-ratio/history', {
      exchange,
      symbol,
      interval,
      limit,
      start_time: startTime,
      end_time: endTime,
    });
  }

  async topLongShortPositionRatioHistory(
    exchange: string,
    symbol: string,
    interval: string,
    limit: number,
    startTime?: number,
    endTime?: number,
  ): Promise<CoinGlassHistoryRow[]> {
    return this.getJsonArray('/api/futures/top-long-short-position-ratio/history', {
      exchange,
      symbol,
      interval,
      limit,
      start_time: startTime,
      end_time: endTime,
    });
  }
}
